### DISCUSSION TITLE
T3100e Fn-key emulation doesn't work

### DISCUSSION THREAD

**Discussion by dcoshea** [86Box/86Box#4816] [Q&A]
### What happened?

Expected behaviour:

1. When in normal DOS in text mode, holding Right Ctrl and pressing the right arrow key should toggle the font between normal and bold.
2. Holding the Right Ctrl key when starting the emulated machine should result in "Check system. Then, press [F1] key....." appearing after the memory check, allowing access to the BIOS settings, or in some other way allow access to the BIOS settings.

I expect this because the Right Ctrl and Right Alt keys are meant to emulate the target machine's Fn key.

Actual behaviour:

1. Nothing happens with that key combination.
2. The machine boots normally despite Right Ctrl being held.

I tried both right arrow keys on my (US 101 key) keyboard, and tried various combinations of Ctrl and Alt keys.

See also #2778 which reported Ctrl + right arrow not working in some older version, and has since been fixed.

### Configuration file

```ini
[General]
vid_renderer = qt_software
host_cpu = Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
uuid = 794cc269-030f-5b95-8cca-58e98efee2e8
scale = 2

[Machine]
machine = t3100e
cpu_family = 286
cpu_speed = 16000000
cpu_multi = 1
cpu_use_dynarec = 0
fpu_softfloat = 0
time_sync = local
mem_size = 1024

[Video]
gfxcard = internal

[Input devices]
mouse_type = msserial

[Sound]
fm_driver = nuked

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = ide_isa
cassette_mode = load

[Microsoft Serial Mouse]
port = 0
buttons = 3

[Hard disks]
hdd_01_parameters = 17, 4, 615, 0, ide
hdd_01_fn = [REDACTED]/hda.img
hdd_01_speed = ramdisk
hdd_01_ide_channel = 0:0

[Floppy and CD-ROM drives]
fdd_01_type = 35_2hd
fdd_02_type = 525_2hd
```

### Operating system

Debian GNU/Linux 12

### CPU

Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz

### 86Box version

v4.2.1 (built from Git tag)

### Build architecture

Linux - x64 (64-bit)

### Build type

- [ ] New recompiler
- [X] Debug build

### Download source

I built 86Box myself (please tell us more about your build configuration)

### Additional context

Build configuration: `--preset dev_debug -D CMAKE_TOOLCHAIN_FILE=[...]/cmake/flags-gcc-x86_64.cmake`

**Comment by OBattler** [comment]
I just committed a tentative fix for it. See if it works.

**Comment by dcoshea** [comment]
> I just committed a tentative fix for it. See if it works.

Thanks, that fixes problem 1 - I can use combinations like Right Ctrl + Right Arrow - but doesn't address problem 2 - holding down Right Ctrl still doesn't enter BIOS setup.

I'm not really sure I understand the code does, and I definitely don't know what the real hardware does, but I imagined that _a_ way to fix this would be to check for the special keys before converting into whatever scan code set the emulated keyboard controller is currently using.  If I understand your fix correctly, it's checking after the emulated keyboard controller has performed scan code translation.  Doesn't that mean that if scan code translation is turned off, and the keyboard controller is using the "wrong" scan code set (not the one corresponding to the values in the `switch` statement), then the Fn keys still won't work?  Feel free to ignore me if this makes no sense!

**Comment by OBattler** [comment]
Yeah, I need a better check for that stuff - perhaps move it to keyboard.c instead.

**Comment by OBattler** [comment]
Also, the thing of holding down the Fn key is not implemented at all. I wonder how that's actually implemented on the real hardware.

**Comment by Fenix770** [comment]
The Toshiba machines T1000/1200/3100e were added by JohnElliot for PCEM. I recall that some PCEM build also broke the Fn key function. All I know is that the three machines have the same Fn key. (for thin/thick font, and enable/disable the external CGA monitor) 

Btw: Also the T1200 seems to support an external MDA card, I don't know if on the real hardware is possible to connect a Hercules card

**Comment by OBattler** [comment]
At this point, this can be closed (or better, converted to a discussion) since the remaining Fn stuff is unknown as to how it was actually implemented in hardware.

**Reply by dcoshea** [reply]
What information is required regarding how it was implemented in hardware?  Information about its behaviour, like whether the real hardware could suffer from some issue where the Fn key didn't work properly depending on the selected scan code set and whether translation was on or off?  Or details about the actual hardware implementation, like the part number of the keyboard controller, and possibly internal details of its implementation?
